### PR TITLE
[documentation] #3509: Generate Kagami CLI help as Markdown

### LINKED ISSUE #3509: [documentation] Auto-generate Kagami docs from clap
### Description

The documentation for Kagami (in its README) is outdated. It will be even more outdated after the merge of #3475.

### Your suggestions

Use [`clap-markdown`](https://crates.io/crates/clap-markdown) or similar tool.

### PR DESCRIPTION
**Pull request by 0x009922**
## Description

_Self-explanatory PR title._

### Linked issue

Closes #3509

### Benefits

Same documentation (from doccomments) for both CLI and GitHub.

### Checklist

- [x] Self-review
- [ ] Merge #3475 
- [ ] Doc review
- [x] Initial code review
- [x] Fix comments
- [ ] Re-review
- [ ] Decide whether we should proceed with `clap-markdown` or with a self-made tool
- [ ] Create a follow-up issue: generate kagami help in CI

### CODE REVIEW

File: tools/kagami/CLI.md
Comment (0x009922) [inline tools/kagami/CLI.md:1]: I don't like the format that `clap-markdown` produces for a couple of reasons:

- Mix of `##` and `#####`
- Using bold in headers: `##### **Options:**`
- Inconsistent spacing
- `↴` character

Not critical though

Comment (0x009922) [reply tools/kagami/CLI.md:1]: Also, it ignores additional text in arguments doccomments and only includes the first line.

**Comment by 0x009922** [comment]
It is not much code to write a better docgen implementation. It might be used for other CLIs as well.

File: tools/kagami/CLI.md
Comment (ilchu) [reply tools/kagami/CLI.md:1]: What do you mean by this? From what I see, the doccomments are included in full (e.g. for genesis or crypto). Or did you mean that it doesn't preserve formatting without the `verbatim` attribute?

File: tools/kagami/CLI.md [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+## `kagami config`
+
+Generate the default client/peer configuration
+
+**Usage:** `kagami config <COMMAND>`
+
+###### **Subcommands:**
+
+* `client` — 
+* `peer` —
```
Comment (ilchu) [inline tools/kagami/CLI.md:221]: Would be good to have short descriptions here as well. Could be done by carrying over the old examples directly into the `Args::Config` docstring, just like you did with `Args::Genesis`, I suppose.

File: tools/kagami/CLI.md
Comment (ilchu) [inline tools/kagami/CLI.md:1]: Also would prefer some different title, as either the old `README.md` or maybe `kagami.md` sound more descriptive to me. And maybe have this markdown generation a part of our CI, just like e.g. schema is (could be a separate small PR though).

File: tools/kagami/CLI.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+**Usage:** `kagami docs`
+
+
+
+## `kagami tokens`
+
+Generate a list of predefined permission tokens and their parameters
+
+**Usage:** `kagami tokens`
```
Comment (ilchu) [inline tools/kagami/CLI.md:249]: The comment about `Args::Config` also applies here, despite this part being insubstantial. Even though it's just two small and obvious examples there, I don't think we lose much by omitting them, but being more explicit never hurts.

File: tools/kagami/src/main.rs [hunk 3]
```diff
@@ -66,18 +160,26 @@ pub enum Args {
...
+    /// issue](https://github.com/hyperledger/iroha/issues/3473)). If you are going to run it outside
     /// of the repo, make sure to pass `--no-default-configuration` flag.
     ///
     /// Be careful with specifying a Dockerhub image as a source: Kagami Swarm only guarantees that
     /// the docker-compose configuration it generates is compatible with the same Git revision it
     /// is built from itself. Therefore, if specified image is not compatible with the version of Swarm
     /// you are running, the generated configuration might not work.
+    #[clap(verbatim_doc_comment)]
     Swarm(swarm::Args),
+    /// Render CLI help message as Markdown
```
Comment (ilchu) [inline tools/kagami/src/main.rs:181]:
```suggestion
    /// Render `kagami` documentation and examples as Markdown
```
As CLI sounds too generic.

**Review by ilchu — CHANGES_REQUESTED** [review]

**Comment by ilchu** [comment]
> It is not much code to write a better docgen implementation. It might be used for other CLIs as well.

Been thinking about this too, you mean ditching the whole `Documented` trait and doing something similar for config endpoint docs?

**Comment by 0x009922** [comment]
> > It is not much code to write a better docgen implementation. It might be used for other CLIs as well.
> 
> Been thinking about this too, you mean ditching the whole `Documented` trait and doing something similar for config endpoint docs?

@ilchu,

Hmm, as far as I understand you are speaking about Iroha configuration documentation, which is generated by `kagami docs`. I am speaking not about it. (Although you are welcome to put your thoughts here: #3507)

Here I mean that `clap-markdown` implementation is not complex, and we can write our own clap-based markdown generator (e.g. `iroha_clap_markdown` crate), with a format that we consider a better option.

File: tools/kagami/CLI.md [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+###### **Options:**
+
+* `--image <IMAGE>` — Use specified docker image
+* `--build <PATH>` — Use local path location of the Iroha source code to build images from
+* `--build-from-github` — Clone `hyperledger/iroha` repo from the revision Kagami is built itself, and use the cloned source code to build images from
+* `-p`, `--peers <PEERS>` — How many peers to generate within the docker-compose
+* `--outdir <OUTDIR>` — Target directory where to place generated files
+* `--outdir-force` — Re-create the target directory if it already exists
+* `--no-default-configuration` — Do not create default configuration in the `<outdir>/config` directory
+* `-s`, `--seed <SEED>` — Might be useful for deterministic key generation
```
Comment (0x009922) [inline tools/kagami/CLI.md:297]: Here is the example of that `clap-markdown` doesn't render additional documentation for arguments. Even if I put `#[clap(verbatim_doc_comment)]`.

For example, `--no-default-configuration`'s full documentation is as follows:

> Do not create default configuration in the `<outdir>/config` directory.
> 
> Default `config.json`, `genesis.json` and `validator.wasm` are generated and put into
> the `<outdir>/config` directory. That directory is specified in the Docker Compose
> `volumes` field.
> 
> If you don't need the defaults, you could set this flag. The `config` directory will be
> created anyway, but you should put the necessary configuration there by yourself.
>

File: tools/kagami/CLI.md
Comment (0x009922) [reply tools/kagami/CLI.md:1]: It definitely should be a part of CI.

Comment (0x009922) [reply tools/kagami/CLI.md:1]: @ilchu, see: https://github.com/hyperledger/iroha/pull/3510#discussion_r1210888850

File: tools/kagami/CLI.md (hunk 4 above)
Comment (ilchu) [reply tools/kagami/CLI.md]: Ah, now I see it, thanks. So it doesn't recurse into the submodules args fully. That should be trivially fixable with a home-brew implementation (and sounds really close to what `Documented` does already anyways).

**Comment by ilchu** [comment]
> @ilchu,
> 
> Hmm, as far as I understand you are speaking about Iroha configuration documentation, which is generated by `kagami docs`. I am speaking not about it. (Although you are welcome to put your thoughts here: #3507)

That's indeed what I was talking about. Then I'll see if I actually can come up with anything actionable up there!

**Comment by QuentinI** [comment]
I'm a bit bothered that running subcommands with `--help` (e.g. `kagami crypto --help`) will output a wall of raw Markdown. I'm not sure the longer docs is even useful there as opposed to keeping it in documentation only; and of course unprocessed Markdown looks quite jarring in a help message

**Review by appetrosyan — APPROVED** [review]

**Comment by 6r1d** [comment]
To summarize our recent conversation, I like the format for the generated docs. Thank you for the detailed demo with Asciinema.

**Comment by 0x009922** [comment]
See: https://github.com/hyperledger/iroha/issues/3509#issuecomment-1581766055

**Pull request closed without merging**